File: 哈希/128.最长连续序列.java
class Solution {
    public int longestConsecutive(int[] nums) {
        if(nums == null || nums.length == 0 ) return 0;
        Set<Integer> set = new HashSet<>(nums.length *2);
        for(int x:nums){
            set.add(x);
        }
        int ans=0;
        for(int x:nums){
            if(set.contains(x-1)) continue;
            int cur=x,len=1;
            while(set.contains(cur+1)){
                cur++;
                len++;
            }
            ans = Math.max(ans,len);
            if(ans*2 >= nums.length){
                break;
            }
        }
        return ans;
    }
}

File: 哈希/128.最长连续序列.go
package hot100

func longestConsecutive(nums []int) int {
	has := map[int]bool{}
	for _, num := range nums {
		has[num] = true
	}
	m := len(has)
	ans := 0
	// 枚举所有数
	for x := range has {
		if _, ok := has[x-1]; ok {
			continue
		}
		cur := x
		cnt := 1
		for has[cur+1] {
			cur++
			cnt++
		}
		ans = max(ans, cnt)
		if ans*2 >= m {
			break
		}
	}
	return ans
}
